RTC_wireless_sensor_nodes_firmware/fermentation_wsn_with_self_calibration/fermentation_wsn_with_self_calibration.ino

```diff
@@ -0,0 +1,389 @@
+  for (int i=0; i<time_between_readings; i++)
+  {
+    dodelay(50000); //1 minute should be 60000 but is not because of variation of internal time source
+    //caution parameter cannot be more than 65000, maybe find better solution
+    //due to internal time source 60000 is longer than 1 minute. so 50s is used.
+  }
+
+} // end of void loop
+
+void calibrate()
+{
+    if(startup==0)
+    {                //if the Arduino just booted up, we need to set some things up first. 
+        digitalWrite(LED,HIGH);  
+        myserial.print("c,0\r");   //take the pH Circuit out of continues mode.
+        // myserial.print("c,1\r");   //enable continuous mode 
+        //myserial.print("L,1\r");      //Turn the LEDs ON
+        //myserial.print("L,0\r");      //Turn the LEDs OFF
+        delay(50);                 //on start up sometimes the first command is missed.
+        digitalWrite(LED,LOW);
+        myserial.print("c,0\r");   //so, let's send it twice.
+        delay(50);                 //a short delay after the pH Circuit was taken out of continues mode is used to make sure we don't over load it with commands.
+        startup=1;                 //startup is completed, let's not do this again during normal operation.
+     }
+      
+   digitalWrite(LED,HIGH);    
+   delay(900);                         //we will take a reading ever 800ms. You can make this much longer or shorter if you like.
+   myserial.print("R\r");             //send it the command to take a single reading.
+  
+   if(string_received==1)  //did we get data back from the ph Circuit?
+   {            
+      ph_calibrate=atof(ph_data);                //Convert string to float 
+     // Serial.println(ph_data);
+      string_received=0;  //reset the string received flag.
+      }             
+     
+     if (ph_calibrate >=1 && ph_calibrate <5) 
+     {
+        if (debug ==1) Serial.print("lowpoint cal. required: ");Serial.println(ph_calibrate);
+        unsigned long current_time = millis()/60000;
+        
+       if(current_time - previousCalibration >= calibrate_after) 
+       {
+          previousCalibration = current_time;
+          lowpoint_cal(); 
+      }
+     }
+     
+     if (ph_calibrate >=5 && ph_calibrate <8) 
+     {
+        if (debug ==1) Serial.print("midpoint cal required: ");Serial.println(ph_calibrate);
+        unsigned long current_time = millis()/60000;
+        
+        if(current_time - previousCalibration >= calibrate_after) 
+        {
+            previousCalibration = current_time;
+            midpoint_cal(); 
+        }
+     }
+  
+     if (ph_calibrate >=8 && ph_calibrate <14)
+     {
+        if (debug ==1) Serial.print("Highpoint cal required: ");Serial.println(ph_calibrate);
+        unsigned long current_time = millis()/60000;
+        if(current_time - previousCalibration >= calibrate_after) 
+        {
+            previousCalibration = current_time;
+            highpoint_cal(); 
+        }
+     }
+} 
+
+void midpoint_cal()
```

Comment:
You can have only one method to do all calculations. Just have a function that does

`ph_calibrate(String level)
`

And simply call it as such
`ph_calibrate('low');`